File: packages/benchmarks/rolodex/scorer.ts
/**
 * Scorer v2 — with relationship type accuracy and full item traces.
 */
import type { Metrics, ItemTrace, RelationshipMetrics, Conversation, Extraction, Resolution, GroundTruthWorld } from './types';

export function m(tp: number, fp: number, fn: number): Metrics {
  const precision = tp + fp > 0 ? tp / (tp + fp) : 1;
  const recall = tp + fn > 0 ? tp / (tp + fn) : 1;
  const f1 = precision + recall > 0 ? (2 * precision * recall) / (precision + recall) : 0;
  return { tp, fp, fn, precision, recall, f1 };
}

function norm(h: string): string { return h.replace(/^@/, '').toLowerCase().trim(); }

export function scoreIdentities(
  conv: Conversation, ext: Extraction
): { metrics: Metrics; items: ItemTrace[] } {
  const exp = conv.expected.identities;
  const act = ext.identities;
  const items: ItemTrace[] = [];
  const matched = new Set<number>();

  for (const a of act) {
    const idx = exp.findIndex((e, i) =>
      !matched.has(i) && e.entityId === a.entityId && e.platform === a.platform && norm(e.handle) === norm(a.handle)
    );
    if (idx >= 0) {
      matched.add(idx);
      items.push({ status: 'TP', label: `${a.entityId}/${a.platform}:${a.handle}`, detail: 'Match' });
    } else {
      items.push({ status: 'FP', label: `${a.entityId}/${a.platform}:${a.handle}`, detail: 'Extra' });
    }
  }
  for (let i = 0; i < exp.length; i++) {
    if (!matched.has(i)) {
      items.push({ status: 'FN', label: `${exp[i].entityId}/${exp[i].platform}:${exp[i].handle}`, detail: 'Missed' });
    }
  }
  return { metrics: m(items.filter(i => i.status === 'TP').length, items.filter(i => i.status === 'FP').length, items.filter(i => i.status === 'FN').length), items };
}

export function scoreRelationships(
  conv: Conversation, ext: Extraction
): { metrics: RelationshipMetrics; items: ItemTrace[] } {
  const exp = conv.expected.relationships;
  const act = ext.relationships;
  const items: ItemTrace[] = [];
  const matched = new Set<number>();
  let typeMatches = 0;
  let totalMatches = 0;

  for (const a of act) {
    const idx = exp.findIndex((e, i) =>
      !matched.has(i) &&
      ((e.entityA === a.entityA && e.entityB === a.entityB) || (e.entityA === a.entityB && e.entityB === a.entityA))
    );
    if (idx >= 0) {
      matched.add(idx);
      totalMatches++;
      const typeOk = exp[idx].type === a.type;
      if (typeOk) typeMatches++;
      items.push({
        status: typeOk ? 'TP' : 'PARTIAL',
        label: `${a.entityA}<->${a.entityB} [${a.type}]`,
        detail: typeOk ? 'Match' : `Pair correct, type wrong (expected: ${exp[idx].type}, got: ${a.type})`,
      });
    } else {
      items.push({ status: 'FP', label: `${a.entityA}<->${a.entityB} [${a.type}]`, detail: 'Extra' });
    }
  }
  for (let i = 0; i < exp.length; i++) {
    if (!matched.has(i)) {
      items.push({ status: 'FN', label: `${exp[i].entityA}<->${exp[i].entityB} [${exp[i].type}]`, detail: 'Missed' });
    }
  }

  const tp = items.filter(i => i.status === 'TP' || i.status === 'PARTIAL').length;
  const fp = items.filter(i => i.status === 'FP').length;
  const fn = items.filter(i => i.status === 'FN').length;
  const typeAccuracy = totalMatches > 0 ? typeMatches / totalMatches : 1;

  return { metrics: { ...m(tp, fp, fn), typeAccuracy }, items };
}

export function scoreTrust(
  conv: Conversation, ext: Extraction
): { metrics: Metrics; items: ItemTrace[] } {
  const exp = conv.expected.trustSignals;
  const act = ext.trustSignals;
  const items: ItemTrace[] = [];
  const matched = new Set<number>();

  for (const a of act) {
    const idx = exp.findIndex((e, i) => !matched.has(i) && e.entityId === a.entityId && e.signal === a.signal);
    if (idx >= 0) {
      matched.add(idx);
      items.push({ status: 'TP', label: `${a.entityId}:${a.signal}`, detail: 'Match' });
    } else {
      items.push({ status: 'FP', label: `${a.entityId}:${a.signal}`, detail: 'Extra' });
    }
  }
  for (let i = 0; i < exp.length; i++) {
    if (!matched.has(i)) {
      items.push({ status: 'FN', label: `${exp[i].entityId}:${exp[i].signal}`, detail: 'Missed' });
    }
  }
  return { metrics: m(items.filter(i => i.status === 'TP').length, items.filter(i => i.status === 'FP').length, items.filter(i => i.status === 'FN').length), items };
}

export function scoreResolution(
  world: GroundTruthWorld, res: Resolution
): { metrics: Metrics; falseMergeRate: number; items: ItemTrace[] } {
  const items: ItemTrace[] = [];
  const matchedLinks = new Set<number>();

  for (const prop of res.links) {
    const isAnti = world.antiLinks.some(al =>
      (al.entityA === prop.entityA && al.entityB === prop.entityB) || (al.entityA === prop.entityB && al.entityB === prop.entityA)
    );
    if (isAnti) {
      items.push({ status: 'FP', label: `${prop.entityA}<->${prop.entityB}`, detail: `FALSE MERGE (anti-link)` });
      continue;
    }
    const idx = world.links.findIndex((link, i) =>
      !matchedLinks.has(i) &&
      ((link.entityA === prop.entityA && link.entityB === prop.entityB) || (link.entityA === prop.entityB && link.entityB === prop.entityA))
    );
    if (idx >= 0) {
      matchedLinks.add(idx);
      items.push({ status: 'TP', label: `${prop.entityA}<->${prop.entityB} (${prop.confidence.toFixed(2)})`, detail: `Signals: ${prop.signals.join('; ')}` });
    } else {
      items.push({ status: 'FP', label: `${prop.entityA}<->${prop.entityB}`, detail: 'No ground truth link' });
    }
  }
  for (let i = 0; i < world.links.length; i++) {
    if (!matchedLinks.has(i)) {
      const link = world.links[i];
      items.push({ status: 'FN', label: `${link.entityA}<->${link.entityB}`, detail: `Missed [${link.difficulty}]: ${link.reason}` });
    }
  }
  const tp = items.filter(i => i.status === 'TP').length;
  const fp = items.filter(i => i.status === 'FP').length;
  const fn = items.filter(i => i.status === 'FN').length;
  return { metrics: m(tp, fp, fn), falseMergeRate: tp + fp > 0 ? fp / (tp + fp) : 0, items };
}

File: packages/benchmarks/rolodex/python_bench/scorer.py
"""Scorer v2 — with relationship type accuracy and full item traces.

Port of the TypeScript scorer.ts.
"""

from __future__ import annotations

import re

from .types import (
    Conversation,
    Extraction,
    GroundTruthWorld,
    ItemTrace,
    Metrics,
    RelationshipMetrics,
    Resolution,
)


def compute_metrics(tp: int, fp: int, fn: int) -> Metrics:
    """Compute precision, recall, and F1 from raw counts."""
    precision = tp / (tp + fp) if (tp + fp) > 0 else 1.0
    recall = tp / (tp + fn) if (tp + fn) > 0 else 1.0
    f1 = (2 * precision * recall) / (precision + recall) if (precision + recall) > 0 else 0.0
    return Metrics(tp=tp, fp=fp, fn=fn, precision=precision, recall=recall, f1=f1)


def _norm(handle: str) -> str:
    """Normalize a handle for comparison: strip @, lowercase, strip whitespace."""
    return re.sub(r"^@", "", handle).lower().strip()


# ── Identity scoring ──────────────────────────────


def score_identities(
    conv: Conversation,
    ext: Extraction,
) -> tuple[Metrics, list[ItemTrace]]:
    """Score extracted identities against expected."""
    expected = conv.expected.identities
    actual = ext.identities
    items: list[ItemTrace] = []
    matched: set[int] = set()

    for a in actual:
        found_idx = -1
        for i, e in enumerate(expected):
            if (
                i not in matched
                and e.entity_id == a.entity_id
                and e.platform == a.platform
                and _norm(e.handle) == _norm(a.handle)
            ):
                found_idx = i
                break
        if found_idx >= 0:
            matched.add(found_idx)
            items.append(ItemTrace(
                status="TP",
                label=f"{a.entity_id}/{a.platform}:{a.handle}",
                detail="Match",
            ))
        else:
            items.append(ItemTrace(
                status="FP",
                label=f"{a.entity_id}/{a.platform}:{a.handle}",
                detail="Extra",
            ))

    for i, e in enumerate(expected):
        if i not in matched:
            items.append(ItemTrace(
                status="FN",
                label=f"{e.entity_id}/{e.platform}:{e.handle}",
                detail="Missed",
            ))

    tp = sum(1 for it in items if it.status == "TP")
    fp = sum(1 for it in items if it.status == "FP")
    fn = sum(1 for it in items if it.status == "FN")
    return compute_metrics(tp, fp, fn), items


# ── Relationship scoring ──────────────────────────


def score_relationships(
    conv: Conversation,
    ext: Extraction,
) -> tuple[RelationshipMetrics, list[ItemTrace]]:
    """Score extracted relationships against expected."""
    expected = conv.expected.relationships
    actual = ext.relationships
    items: list[ItemTrace] = []
    matched: set[int] = set()
    type_matches = 0
    total_matches = 0

    for a in actual:
        found_idx = -1
        for i, e in enumerate(expected):
            if i not in matched and (
                (e.entity_a == a.entity_a and e.entity_b == a.entity_b)
                or (e.entity_a == a.entity_b and e.entity_b == a.entity_a)
            ):
                found_idx = i
                break
        if found_idx >= 0:
            matched.add(found_idx)
            total_matches += 1
            type_ok = expected[found_idx].type == a.type
            if type_ok:
                type_matches += 1
            items.append(ItemTrace(
                status="TP" if type_ok else "PARTIAL",
                label=f"{a.entity_a}<->{a.entity_b} [{a.type}]",
                detail="Match" if type_ok else (
                    f"Pair correct, type wrong "
                    f"(expected: {expected[found_idx].type}, got: {a.type})"
                ),
            ))
        else:
            items.append(ItemTrace(
                status="FP",
                label=f"{a.entity_a}<->{a.entity_b} [{a.type}]",
                detail="Extra",
            ))

    for i, e in enumerate(expected):
        if i not in matched:
            items.append(ItemTrace(
                status="FN",
                label=f"{e.entity_a}<->{e.entity_b} [{e.type}]",
                detail="Missed",
            ))

    tp = sum(1 for it in items if it.status in ("TP", "PARTIAL"))
    fp = sum(1 for it in items if it.status == "FP")
    fn = sum(1 for it in items if it.status == "FN")
    type_accuracy = type_matches / total_matches if total_matches > 0 else 1.0

    base = compute_metrics(tp, fp, fn)
    return (
        RelationshipMetrics(
            tp=base.tp,
            fp=base.fp,
            fn=base.fn,
            precision=base.precision,
            recall=base.recall,
            f1=base.f1,
            type_accuracy=type_accuracy,
        ),
        items,
    )


# ── Trust signal scoring ──────────────────────────


def score_trust(
    conv: Conversation,
    ext: Extraction,
) -> tuple[Metrics, list[ItemTrace]]:
    """Score extracted trust signals against expected."""
    expected = conv.expected.trust_signals
    actual = ext.trust_signals
    items: list[ItemTrace] = []
    matched: set[int] = set()

    for a in actual:
        found_idx = -1
        for i, e in enumerate(expected):
            if i not in matched and e.entity_id == a.entity_id and e.signal == a.signal:
                found_idx = i
                break
        if found_idx >= 0:
            matched.add(found_idx)
            items.append(ItemTrace(
                status="TP",
                label=f"{a.entity_id}:{a.signal}",
                detail="Match",
            ))
        else:
            items.append(ItemTrace(
                status="FP",
                label=f"{a.entity_id}:{a.signal}",
                detail="Extra",
            ))

    for i, e in enumerate(expected):
        if i not in matched:
            items.append(ItemTrace(
                status="FN",
                label=f"{e.entity_id}:{e.signal}",
                detail="Missed",
            ))

    tp = sum(1 for it in items if it.status == "TP")
    fp = sum(1 for it in items if it.status == "FP")
    fn = sum(1 for it in items if it.status == "FN")
    return compute_metrics(tp, fp, fn), items


# ── Resolution scoring ────────────────────────────


def score_resolution(
    world: GroundTruthWorld,
    res: Resolution,
) -> tuple[Metrics, float, list[ItemTrace]]:
    """Score entity resolution.

    Returns ``(metrics, false_merge_rate, items)``.
    """
    items: list[ItemTrace] = []
    matched_links: set[int] = set()

    for prop in res.links:
        # Check anti-links
        is_anti = any(
            (al.entity_a == prop.entity_a and al.entity_b == prop.entity_b)
            or (al.entity_a == prop.entity_b and al.entity_b == prop.entity_a)
            for al in world.anti_links
        )
        if is_anti:
            items.append(ItemTrace(
                status="FP",
                label=f"{prop.entity_a}<->{prop.entity_b}",
                detail="FALSE MERGE (anti-link)",
            ))
            continue

        found_idx = -1
        for i, link in enumerate(world.links):
            if i not in matched_links and (
                (link.entity_a == prop.entity_a and link.entity_b == prop.entity_b)
                or (link.entity_a == prop.entity_b and link.entity_b == prop.entity_a)
            ):
                found_idx = i
                break
        if found_idx >= 0:
            matched_links.add(found_idx)
            items.append(ItemTrace(
                status="TP",
                label=f"{prop.entity_a}<->{prop.entity_b} ({prop.confidence:.2f})",
                detail=f"Signals: {'; '.join(prop.signals)}",
            ))
        else:
            items.append(ItemTrace(
                status="FP",
                label=f"{prop.entity_a}<->{prop.entity_b}",
                detail="No ground truth link",
            ))

    for i, link in enumerate(world.links):
        if i not in matched_links:
            items.append(ItemTrace(
                status="FN",
                label=f"{link.entity_a}<->{link.entity_b}",
                detail=f"Missed [{link.difficulty}]: {link.reason}",
            ))

    tp = sum(1 for it in items if it.status == "TP")
    fp = sum(1 for it in items if it.status == "FP")
    fn = sum(1 for it in items if it.status == "FN")
    false_merge_rate = fp / (tp + fp) if (tp + fp) > 0 else 0.0

    return compute_metrics(tp, fp, fn), false_merge_rate, items
